modAL/datamanager.py

```diff
@@ -0,0 +1,136 @@
+        where UD and LD are subsets of OD.
+        The active learner operates on UD and returns indexes relative to it. We want to store those indices with respect
+        to OD, and sometimes see the subset of labels of LD. (The subset of labels of UD is Null)
+
+        That's a fancy way of saying there is a lot book keeping to be done and this class solves that by doing it for you
+
+        The main idea is that we store a mask (labeeld_mask) of indices that have been labeled and then expose UD , LD
+        and the labels by using fancy indexing with that mask. The manager exposes a an add_labels method which lets the
+        user add labels indexed with respect to UD and it will adjust the indices so that they match OD.
+
+        :param features: An array of the features that will be used for AL.
+        :param labels: Any prexesiting labels. Each label is a tuple(idx,label)
+        :param source: A list of the original data
+        """
+        self.features = features
+
+        self._labels =np.empty(shape=self.features.shape[0],dtype=labels_dtype)
+        self.labeled_mask = np.zeros(self.features.shape[0], dtype=bool)
+        self.sources = np.array(sources if sources else [])
+
+    @property
+    def labels(self):
+        '''
+
+        Returns the labels indexed with respect to LD
+
+        '''
+        return self._labels[self.labeled_mask]
+
+    @property
+    def unlabeled_mask(self):
+        '''
+
+        Returns: a mask which is true for all unlabeled points
+
+        '''
+        return np.logical_not(self.labeled_mask)
+
+    def _update_masks(self, labels: Union[LabelList, Label]):
+        for label in labels:
+            self.labeled_mask[label[0]] = True
+
+    def _offset_new_labes(self, labels: LabelList):
+        """
+        This is where the magic happens.
+        We take self.unlabeled_mask.nonzero()[0] which gives us an array of the indices that appear in the unlabeled
+        data. So if the original label was at position 0 we look up the "real index" in the unlabeled_indices array to
+        get it's true index
+        :param labels:
+        :return:
+        """
+        if len(self._labels) == 0:
+            # Nothing to correct in this case
+            return labels
+        correctLabels: LabelList = []
+        unlabeled_indices = self.unlabeled_mask.nonzero()[0]
+
+        for label in labels:
+            newIndex = unlabeled_indices[label[0]]
+            newLabel: Label = (newIndex, label[1])
+            correctLabels.append(newLabel)
+        return correctLabels
+
+    def add_labels(self, labels: LabelList,offset_to_unlabeled=True):
+        if isinstance(labels, tuple):  # if this is a single example
+            labels: LabelList = [labels]
+        elif isinstance(labels, list):
+            pass
+        else:
+            raise Exception(
+                "Malformed input. Please add either a tuple (ix,label) or a list [(ix,label),..]"
+            )
```

Comment:
This should be a `TypeError` instead of a generic exception, since it perfectly fits the scope of `TypeError`.